[Inductor] Move `GPU_TYPE`(The runtime avaliable gpu type, cuda or xpu) from

etaf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132843
* #132748
* __->__ #132740

Move GPU_TYPE(The runtime avaliable gpu type, cuda or xpu) from `testing/_internal/inductor_utils.py` to `_inductor/utils.py`. So that we can use it in Inductor, not limited in test case.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

etaf committed:
[Inductor] Move `GPU_TYPE`(The runtime avaliable gpu type, cuda or xpu) from
testing/_internal/inductor_utils.py to _inductor/utils.py. So that we
can use it in Inductor, not limited in test case.

[ghstack-poisoned]

malfet commented on torch/_inductor/utils.py:
```diff
@@ -44,6 +44,20 @@
 import torch
+
+
+GPU_TYPES = ["cuda", "xpu"]
+
+
+def _get_gpu_type():
+    avail_gpus = [x for x in GPU_TYPES if getattr(torch, x).is_available()]
+    assert len(avail_gpus) <= 1
+    gpu_type = "cuda" if len(avail_gpus) == 0 else avail_gpus.pop()
+    return gpu_type
+
+
+GPU_TYPE = _get_gpu_type()
```
Can you please move this code below the `from XYZ import ...` block? (may be towards the end of the file, unless other utils depend on it

etaf replied:
Thank you, your comments helps a lot. In this place, since I need to use _get_gpu_type() to generalize the device-bias code inside dynamo in my later PR, I need to import _get_gpu_type inside dynamo, and in order to avoid the loop of importing, I just put this function in this file before the import of dynamo. Do you have any suggestions for this scenario? Or do I need to put it outside the inductor?

etaf replied:
@malfet can you help review the changes?

etaf replied:
> Can you please move this code below the from XYZ import ... block? (may be towards the end of the file, unless other utils depend on it

Hi, @malfet , do you have any comment for this change?

malfet commented on torch/_inductor/utils.py:
```diff
@@ -44,6 +44,20 @@
 import torch
+
+
+GPU_TYPES = ["cuda", "xpu"]
+
+
+def _get_gpu_type():
+    avail_gpus = [x for x in GPU_TYPES if getattr(torch, x).is_available()]
+    assert len(avail_gpus) <= 1
+    gpu_type = "cuda" if len(avail_gpus) == 0 else avail_gpus.pop()
+    return gpu_type
+
+
+GPU_TYPE = _get_gpu_type()
```
Please don't create global `GPU_TYPE` environment variable, but rather memoize results of `_get_gpu_type()`, otherwise importing inductor would cause GPU subsystem initialization even if user only plans to use inductor for CPU 
```suggestion
GPU_TYPES = ["cuda", "xpu"]


@functools.lru_cache(None)
def _get_gpu_type():
    avail_gpus = [x for x in GPU_TYPES if getattr(torch, x).is_available()]
    assert len(avail_gpus) <= 1
    gpu_type = "cuda" if len(avail_gpus) == 0 else avail_gpus.pop()
    return gpu_type


```

etaf replied:
> Please don't create global `GPU_TYPE` environment variable, but rather memoize results of `_get_gpu_type()`, otherwise importing inductor would cause GPU subsystem initialization even if user only plans to use inductor for CPU
Thanks for the reminder.

malfet reviewed (changes requested):
Please do not introduce any code in `_inductor` that initializes GPU subsystem as soon as it's imported (I.e. global var like `IS_CUDA_AVAILABLE = torch.cuda.is_available()` ) 

Instead, please use functions, that will delay the initialization from import time to the call time

jansel reviewed (changes requested):
+1 to @malfet's comment

`import torch._inductor` should not have side effects

etaf committed:
Update on "[Inductor] Move `GPU_TYPE`(The runtime avaliable gpu type, cuda or xpu) from"


Move GPU_TYPE(The runtime avaliable gpu type, cuda or xpu) from `testing/_internal/inductor_utils.py` to `_inductor/utils.py`. So that we can use it in Inductor, not limited in test case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

etaf committed:
Update on "[Inductor] Move `GPU_TYPE`(The runtime avaliable gpu type, cuda or xpu) from"


Move GPU_TYPE(The runtime avaliable gpu type, cuda or xpu) from `testing/_internal/inductor_utils.py` to `_inductor/utils.py`. So that we can use it in Inductor, not limited in test case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jansel commented on torch/testing/_internal/inductor_utils.py:
```diff
@@ -41,18 +42,13 @@ def test_cpu():
 HAS_GPU = HAS_CUDA or HAS_XPU
 
-GPUS = ["cuda", "xpu"]
+GPU_TYPE = get_gpu_type()
```
Let's refactor this so it doesn't initialize the GPU subsystem on import.

etaf replied:
Hi @jansel, HAS_CUDA/HAS_XPU/HAS_GPU are already in the header of this file, and in Inductor UT it is imported to get HAS_CUDA/HAS_XPU/HAS_GPU. Seems onece import `torch.testing._internal.inductor_utils`, the GPU subsystem is always initialized.

etaf commented:
@pytorchbot merge
